src/program/latex/mod.rs

```diff
@@ -0,0 +1,1600 @@
+/// and footer while the body contains content and controls its presentation.
+struct Document {
+    header: String,
+    body: String,
+    footer: String,
+}
+
+impl Default for Document {
+    fn default() -> Self {
+        Self {
+            header: r"\documentclass[convert={density=300,outext=.png}]{standalone}
+\usepackage[margin=1in]{geometry}
+\usepackage{tikz}
+\usetikzlibrary{quantikz}
+\begin{document}
+\begin{tikzcd}"
+                .to_string(),
+            body: "".to_string(),
+            footer: r"\end{tikzcd}
+\end{document}"
+                .to_string(),
+        }
+    }
+}
+
+impl Display for Document {
+    /// Returns the entire document in LaTeX string.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}{}{}", self.header, self.body, self.footer)
+    }
+}
+
+/// A Diagram represents a collection of wires in a circuit. It encodes the
+/// wires row in the diagram and its relationship to other wires. A row is one
+/// of the wires in the circuit BTreeMap. Diagram tracks relationships between
+/// wires with two pieces of information--1. the wires row (its order in the
+/// BTreeMap), and 2. the column that spreads between all wires that pass
+/// through a multi qubit gate 'e.g. CNOT'. The size of the diagram can be
+/// measured by multiplying the column with the length of the circuit. This is
+/// an [m x n] matrix where each element in the matrix represents an item to be
+/// rendered onto the diagram using one of the [`Quantikz`] commands.
+#[derive(Debug)]
+struct Diagram {
+    /// customizes how the diagram renders the circuit
+    settings: Settings,
+    /// total number of elements on each wire
+    column: u32,
+    /// column at which qubits in positional order form relationships
+    relationships: HashMap<u32, Vec<u64>>,
+    /// a BTreeMap of wires with the name of the wire as the key
+    circuit: BTreeMap<u64, Box<Wire>>,
+}
+
+impl Default for Diagram {
+    fn default() -> Self {
+        Self {
+            settings: Settings::default(),
+            column: 0,
+            relationships: HashMap::new(),
+            circuit: BTreeMap::new(),
+        }
+    }
+}
+
+impl Diagram {
+    /// Compares qubits from a single instruction associated with a column on
+    /// the circuit to all of the qubits used in the quil program. If a qubit
+    /// from the quil program is not found in the qubits in the single
+    /// instruction line, then an empty slot is added to that column on the
+    /// qubit wire of the circuit.
+    ///
+    /// # Arguments
+    /// `&mut self` - exposes the Circuit
+    /// `qubits` - qubits used in the quil program
+    /// `instruction` - exposes qubits in a single instruction
+    fn set_qw(&mut self, qubits: &HashSet<Qubit>, instruction: &Instruction) {
+        for program_qubit in qubits {
+            let mut found = false;
+            match &instruction {
```

Comment:
Two things:

(1) If this function takes a single `Instruction`, and that `Instruction` has to be a `Gate`, then why bother checking for every qubit in `qubits`? Would it make more sense to do the match first, and have the loop in the body?

(2) The `HashSet` API has a [`contains(other)`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.contains) method. Why not use that instead of managing the found boolean and nesting loops? Something like:

```rs
match &instruction {
    instruction::Instruction::Gate(gate) => {
        for gate_qubit in &gate.qubits {
            if !qubits.contains(gate_qubit) {
                match program_qubit {
                    Instruction::Qubit::Fixed(q) => {
                        self.circuit
                            .get_mut(&q)
                            .and_then(|wire| wire.empty.insert(self.column, Command::Qw));
                    },
                    _ => (),
                }
            }
        }
    },
    _ => (),
}